torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -378,6 +378,12 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::allreduce(
     gpuGuard.set_index(devices[i].index());
     at::cuda::CUDAStream& ncclStream = ncclStreams_[key][i];
 
+    // Input `tensors` are created on a worker stream and used in a different
+    // ncclStream. Hence, `tensors` must record the ncclStream to prevent being
+    // freed before ncclAllReduce finishes.
+    c10::cuda::CUDACachingAllocator::recordStream(
+      tensors[i].data_ptr(), ncclStream);
+
```

Comment:
You should move this to _after_ `ncclGroupEnd` to ensure the ordering is correct. This call will record an event to the NCCL stream, and at this location, the NCCL kernel has not been launched yet.

Reply:
IIUC, this call [adds](https://github.com/pytorch/pytorch/blob/master/c10/cuda/CUDACachingAllocator.cpp#L389) the `ncclStream` to a set, and only records an event when the worker stream [attempts to delete the tensor](https://github.com/pytorch/pytorch/blob/master/c10/cuda/CUDACachingAllocator.cpp#L319). In this case, the input `tensors` will not be deleted on the worker stream throughout the `allReduce` invocation. So, it should be OK?

Reply:
I see, thanks for the clarification! My understand was that this would record an event, but since it stores the stream the location where this is done doesn't matter.